testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+Both of these problems were eventually solve by using the GitHub API in combination with the information we get from the workflow event's context. Getting the commit reference of the pull request that triggered the workflow is the main challenge, when we have that, creating statuses on commits is trivial. 
+
+##### What branch am i even in?
+
+The 'context' a workflow runs in is the branch that actions/checkout will checkout by default. In this case that is `main`. We don't want `main`.
+
+Figuring out what branch or pull request triggered a workflow run is surprisingly difficult depending on the event that you allow to trigger it. The data you have access to in a ` workflow_run` event is pretty limited. It is okay for pull requests, you get the PR number, but for pushes and other events it can be challenging. We trigger `workflow_runs` from the following events:
+
+- `pull_request` - this is fine, we get the PR number.
+- `push` - not fine, but we get the commit SHA which will do.
+- `issue_comment` - couldn't be less fine, we only get the title.
+
+*The reason we use the `issue_comment` event is because pull request are actually a special type of issue when it comes to the title, body, and replies.*
+
+It is much easier to find a SHA from a PR number than the other way around but both are possible, getting a PR from an issue title, or PR title is pretty error-prone. We typically need the following information in our workflows:
+
+- Pull request number, if one exists (we create comments in a the PR for certain jobs).
+- Source repository
+- Source branch
+- The HEAD SHA of the source branch (sometimes we want to check this out, we always want checks to appear on this ref)
+- The SHA of the magical 'merge' branch that github creates (we want to check this out usually)
+- Any labels for the PR (we use these for certain conditional jobs)
+
+<details>
+<summary>A magical branch, you say?</summary>
+
+GitHub actually creates two magical refs. `pull/<pr-number>/head` and `pull/<pr-number>/merge`. Both of these refs are read-only, you cannot push to them no matter how many `-f`s you add.
+
+The `head` variant is pretty much the same as the HEAD of the pr branch, except it exists in the target repo regardless of whether it was created from a fork or not. This is quite nice as the commit SHA for this ref will be the same as the commit SHA for the HEAD of the source branch. This makes checking out the branch easier.
+
+The `merge` variant is special. This is a branch that merges the PR changes into the target branch. `pull_request` events have this branch set as their 'default' and it is what gets checked out by default in `pull_request` workflows. The beauty of this branch is that any tests you run against it are essentially being run on the merged result of this PR and `main`. This isn't commonly know but it is exactly what you want in a pull request.
```

Comment:
I'm pretty sure it is but it might not always be true. I'd need to check but it is possible that `pull/<pr>/merge` is preferred but it falls back to `pull/<pr>/head`. They would know ahead of time if it is possible to run checks on the former because it is created every time a new commit is pushed prior to running checks.

If there are no conflicts, it definitely uses the merged result though, that I have tested.